src/io/read_tree_genesis_hdf5.c

```diff
@@ -0,0 +1,1137 @@
+            forestnr);
+
+    int64_t offset=0;
+    const int start_snap = gen->min_snapnum;
+    const int end_snap = gen->min_snapnum + gen->maxsnaps - 1;//maxsnaps already includes a +1,
+    for(int isnap=end_snap;isnap>=start_snap;isnap--) {
+        if(offset > INT_MAX) {
+            fprintf(stderr,"Error: In function %s> Can not correctly represent %"PRId64" as an offset in the 32-bit variable within "
+                    "the LHaloTree struct. \n", __FUNCTION__, offset);
+            ABORT(INTEGER_32BIT_TOO_SMALL);
+        }
+        char dset_name[MAX_STRING_LEN + 1];
+        snprintf(dset_name, MAX_STRING_LEN, "ForestInfo/Snaps/Snap_%03d", isnap);
+        READ_GENESIS_PARTIAL_FORESTINFO(gen->meta_fd, dset_name, "NumHalosInForest", &h5_global_forestnum, &h5_single_item, &(nhalos_per_snap[isnap]));
+
+        forest_local_offsets[isnap] = offset;
+        /* fprintf(stderr,"forest_local_offsets[%03d] = %d\n", isnap, forest_local_offsets[isnap]); */
+        offset += nhalos_per_snap[isnap];
+    }
+
+    /* Check that the number of halos to read in agrees with that derived with the per snapshot one */
+    if(offset != nhalos) {
+        fprintf(stderr,"Error: On ThisTask = %d while processing task-local-forestnr = %"PRId64" "
+                " file-local-forestnr = %"PRId64"and global forestnum = %"PRId64" located in the file = %d\n",
+                run_params->ThisTask, forestnr, forests_info->original_treenr[forestnr], forestnum_across_all_files, filenum);
+        fprintf(stderr,"Expected the 'nhalos_per_snap' array to sum up to 'nhalos' but that is not the case\n");
+        fprintf(stderr,"Sum(nhalos_per_snap) = %"PRId64" nhalos = %"PRId64"\n", offset, nhalos);
+        fprintf(stderr,"Now printing out individual values of the nhalos_per_snap\n");
+        for(int isnap=end_snap;isnap>=start_snap;isnap--) {
+            fprintf(stderr,"nhalos_per_snap[%03d] = %09"PRId64"\n", isnap, nhalos_per_snap[isnap]);
+        }
+        fprintf(stderr,"Now printing out the offset need per file\n");
+        for(int i=0;i<gen->totnfiles;i++) {
+            fprintf(stderr,"gen->offset_for_global_forestnum[%04d] = %09"PRId64"\n", i, gen->offset_for_global_forestnum[i]);
+        }
+
+        ABORT(INVALID_VALUE_READ_FROM_FILE);
+    }
+
+    *halos = mymalloc(sizeof(struct halo_data) * nhalos);//the malloc failure check is done within mymalloc
+    struct halo_data *local_halos = *halos;
+    for(int64_t i=0;i<nhalos;i++) {
+        local_halos[i].FirstHaloInFOFgroup = nhalos + 1;
+        local_halos[i].NextHaloInFOFgroup = -1;
+
+        local_halos[i].FirstProgenitor = -1;
+        local_halos[i].NextProgenitor = -1;
+        local_halos[i].Descendant = -1;
+    }
+
+    // The max. size of the data to be read in would be "nhalos" * NDIM (== 3 for pos/vel) * sizeof(double)
+    void *buffer = calloc(nhalos * NDIM * sizeof(double), sizeof(*buffer));
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['*buffer' is of type 'void', the behaviour of 'sizeof(void)' is not covered by the ISO C standard.](https://app.codacy.com/manual/jacobseiler/sage-model/pullRequest?prid=4620074)